automap/src/comm_layer/pmp.rs

```diff
@@ -1465,6 +1882,321 @@ mod tests {
+            &Some(MappingConfig {
+                hole_port: 2222,
+                next_lifetime: Duration::from_secs(10),
+                remap_interval: Duration::from_secs(0),
+            }),
+        );
+
+        // No complaint from MappingAdderMock about not having a result for add_mapping; test passes
+    }
+
+    #[test]
+    fn handle_announcement_rejects_temporarily_unsuccessful_result_code() {
+        init_test_logging();
+        let router_address = SocketAddr::from_str("7.7.7.7:1234").unwrap();
+        let mut packet = PmpPacket::default();
+        packet.direction = Direction::Response;
+        packet.opcode = Opcode::MapTcp;
+        packet.result_code_opt = Some(ResultCode::OutOfResources);
+        packet.opcode_data = Box::new(MapOpcodeData::default());
+        let mut buffer = [0u8; 100];
+        let buflen = packet.marshal(&mut buffer).unwrap();
+        let socket = UdpSocketWrapperMock::new()
+            .set_read_timeout_result(Ok(()))
+            .send_to_result(Ok(100))
+            .recv_from_result(Ok((buflen, router_address)), buffer[0..buflen].to_vec());
+        let factories = Factories {
+            socket_factory: Box::new(UdpSocketWrapperFactoryMock::new().make_result(Ok(socket))),
+            free_port_factory: Box::new(FreePortFactoryMock::new().make_result(1234)),
+        };
+        let change_handler_log_arc = Arc::new(Mutex::new(vec![]));
+        let change_handler_log_inner = change_handler_log_arc.clone();
+        let change_handler: ChangeHandler =
+            Box::new(move |change| change_handler_log_inner.lock().unwrap().push(change));
+        let logger = Logger::new("test");
+        let transactor = PmpTransactor::new();
+        let mut subject = ThreadGuts::new(
+            &transactor,
+            router_address.ip(),
+            Box::new(UdpSocketWrapperMock::new()),
+            change_handler,
+            unbounded().1,
+        );
+        subject.router_addr = router_address;
+        subject.factories_arc = Arc::new(Mutex::new(factories));
+        subject.mapping_adder_arc = Arc::new(Mutex::new(Box::new(MappingAdderMock::new())));
+        subject.logger = logger;
+
+        subject.handle_announcement(
+            Ipv4Addr::from_str("4.3.2.1").unwrap(),
+            &mut Some(MappingConfig {
+                hole_port: 2222,
+                next_lifetime: Duration::from_secs(10),
+                remap_interval: Duration::from_secs(0),
+            }),
+        );
+
+        let change_handler_log = change_handler_log_arc.lock().unwrap();
+        let err_msg = "Remapping after IP change failed; Node is useless: OutOfResources";
+        assert_eq!(
+            *change_handler_log,
+            vec![AutomapChange::Error(AutomapError::TemporaryMappingError(
+                err_msg.to_string()
+            ))]
+        );
+        TestLogHandler::new().exists_log_containing(&format!("ERROR: test: {}", err_msg));
+    }
+
```

Comment:
The previous and the next test are almost identical, please create an assert function which could be common for both tests and with just very few arguments: result code, expected_err_msg and maybe also test name (because I don't think necessarily it must be truth, by general principle, that the messages will never meet similar "peers" emitted by other tests in this file; so I'd be more careful there).

You can also consider a test body that even the test before these two would reuse. There is just a small difference again...

Reply:
Wellll...okay.